olad/NotifySystemd.cpp

```diff
@@ -0,0 +1,51 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * NotifySystemd.cpp
+ * Provides wrapped access to the systemd notification interface.
+ * Copyright (C) 2018 Shenghao Yang
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif  // HAVE_CONFIG_H
+
+#if HAVE_LIBSYSTEMD
+#include <systemd/sd-daemon.h>
+#include <errno.h>
+#include <string.h>
+#endif  // HAVE_LIBSYSTEMD
+
+#include "ola/Logging.h"
+
+#include "olad/Systemd.h"
+
+namespace ola {
+
+int notify_systemd(int unset_environment, const char *state) {
+  int rtn = sd_notify(unset_environment, state);
+  if (rtn < 0) {
+    char buf[1024];
+    OLA_WARN << "Error sending notification to systemd: " <<
+      strerror_r(-rtn, buf, sizeof(buf));
```

Comment:
I think you'll only print the error code from strerror_r currently. Looking at https://linux.die.net/man/3/strerror_r it seems you should be able to just do `<< strerror(-rtn)`.

Can you also move the second << to the next line and align with the one above.

Reply:
Also I think you should add a configure check for strerror_r (see https://github.com/OpenLightingProject/ola/blob/master/configure.ac#L185 and e.g. https://github.com/OpenLightingProject/ola/blob/master/common/base/Credentials.cpp#L342 ), it's the first time it's being used in the codebase and we run on a variety of weird platforms. See also some similar fun and discussion with readdir_r https://github.com/OpenLightingProject/ola/issues/1055 .

Reply:
Ah yes, in which case can we just have a comment for our future selves if we switch away from the GNU extensions in future.

Well currently multiple plugins call strerror (and some do indeed spawn their own threads, at least one call is in a spawned thread), so we may have that issue already.

We should probably be moving to strerror_r throughout, but equally I wouldn't want this one usage to break the build on machines that don't have it, or by adding it to configure stop it building on those OSes.

Reply:
> Ah yes, in which case can we just have a comment for our future selves if we switch away from the GNU extensions in future.

Sure, alright.

> We should probably be moving to strerror_r throughout, but equally I wouldn't want this one usage to break the build on machines that don't have it, or by adding it to configure stop it building on those OSes.

Any modern system with `systemd` should have this - deliberately using `sterrror()` feels pretty icky. Maybe we can take a page from `nginx`: they call `strerror()` in a section where it is known that no other thread can call `strerror()` and caches that for use.

Have a look at their implementation: https://github.com/nginx/nginx/blob/f8a9d528df92c7634088e575e5c3d63a1d4ab8ea/src/os/unix/ngx_errno.c

Reply:
Fixed in https://github.com/OpenLightingProject/ola/pull/1444/commits/469361d2b1c52d0475621cb97653511fa7d60201